framework/source/class/qx/io/DynamicScriptLoader.js

```diff
@@ -0,0 +1,245 @@
+     *     Returns the name of the last script you can expect to be loaded if all scripts are already loaded null will be returned.
+     *
+     */
+    load: function(codeArr, callback, context) {
+      var lastScript = null;
+      codeArr.forEach(function(script) {
+        if (this.__ADDED[script] !== true) {
+          this.__QUEUE.push(script);
+          lastScript = script;
+          this.__ADDED[script] = true;
+        }
+      }, this);
+      if (callback) {
+        var cb = new qx.util.DeferredCall(callback, context);
+        if (lastScript !== null) {
+          var id = this.addListener('loaded', function(e) {
+            var data = e.getData();
+            // this.debug("Loading " + data.script + " completed");
+            if (data.script == lastScript) {
+              this.removeListenerById(id);
+              cb.schedule();
+            }
+          }, this);
+        }
+        else {
+          cb.schedule();
+        }
+      }
+      this.__loadScripts();
+      return lastScript;
+    },
+
+
+    /**
+     * Chain loading scripts.
+     *
+     * Recursively called until the array of scripts is consumed
+     *
+     * @param codeArr {Array} an array with the uri names of the scripts
+     */
+    __loadScripts: function() {
+
+      if (this.__LOADING === true) {
+        return;
+      }
+
+      var script = this.__QUEUE.shift();
+
+      if (script) {
+        this.__LOADING = true;
+        var uri = qx.util.ResourceManager.getInstance().toUri(script);
+
+        var loader = new qx.bom.request.Script();
+
+        loader.on("load", function(request) {
+          this.__LOADED[script] = true;
+          this.__LOADING = false;
+          this.fireDataEvent('loaded', {
+            script: script,
+            uri: uri,
+            status: request.status
+          });
+          // start the next load event after the current thread has ended
+          var call = new qx.util.DeferredCall(this.__loadScripts, this);
+          call.schedule();
+        }, this);
+
+        loader.on("error", function(request) {
+          this.__LOADING = false;
+          this.fireDataEvent('failed', {
+            script: script,
+            uri: uri,
+            status: request.status
+          });
+        }, this);
+
+        loader.on("timeout", function(request) {
+          this.__LOADING = false;
+          this.fireDataEvent('failed', {
+            script: script,
+            uri: uri,
+            status: request.status
+          });
+        }, this);
+        // this.debug("Loading " + script + " started");
+        loader.open("GET", uri);
+        loader.send();
+      }
+      // no more scripts in codeArr, which means that all the
+      // scripts were loaded.
+      else {
+        // even when we 'fall through' make sure the event gets fired AFTER the current
+        // code is done.
+        var ready = new qx.util.DeferredCall(function(){this.fireEvent("ready")},this);
+        ready.schedule();
```

Comment:
And there's no need to defer calling the callback here, either, since we don't have any more scripts to start loading.